java/ql/src/semmle/code/java/security/CommandArguments.qll

```diff
@@ -0,0 +1,217 @@
+      a.getIndexExpr() instanceof CompileTimeConstantExpr
+    )
+  }
+
+  /** Gets an expression that is written to the given index of this array at he given use. */
+  Expr getAWrite(int index, RValue use) {
+    exists(Assignment a, ArrayAccess acc |
+      acc.getArray() = use and
+      use = this.getAUse() and
+      index = acc.getIndexExpr().(CompileTimeConstantExpr).getIntValue() and
+      acc = a.getDest() and
+      result = a.getRhs()
+    )
+  }
+
+  /** Gets an expression that is written to the given index of this array. */
+  Expr getAWrite(int index) { result = getAWrite(index, _) }
+
+  predicate isNotShell() {
+    exists(Expr e | e = this.(CommandArgArrayImmutableFirst).getFirstElement() | not isShell(e))
+  }
+}
+
+/**
+ * A `CommandArgArray` whose element at index 0 is never written to, except possibly once to initialise it.
+ */
+private class CommandArgArrayImmutableFirst extends CommandArgumentArray {
+  CommandArgArrayImmutableFirst() {
+    this.getDefiningExpr() instanceof ImmutableFirstArrayExpr and
+    forall(RValue use | exists(this.getAWrite(0, use)) | use = this.getAFirstUse())
+  }
+
+  /** Gets the first element of this array. */
+  Expr getFirstElement() {
+    result = getAWrite(0)
+    or
+    not exists(getAWrite(0)) and
+    result = getDefiningExpr().(ImmutableFirstArrayExpr).getFirstElement()
+  }
+}
+
+/**
+ * An expression that evaluates to an array of strings whose first element is immutable.
+ */
+private class ImmutableFirstArrayExpr extends Expr {
+  ImmutableFirstArrayExpr() {
+    this.getType() instanceof ArrayOfStringType and
+    (
+      this.(Assignment).getRhs() instanceof ImmutableFirstArrayExpr
+      or
+      this.(LocalVariableDeclExpr).getInit() instanceof ImmutableFirstArrayExpr
+      or
+      this instanceof ArrayInit
+      or
+      this instanceof ArrayCreationExpr
+      or
+      this.(RValue) = any(CommandArgArrayImmutableFirst caa).getAUse()
+      or
+      exists(MethodAccess ma, Method m |
+        ma.getMethod() = m and
+        m.getDeclaringType().hasQualifiedName("java.util", "Arrays") and
+        m.hasName("copyOf") and
+        ma.getArgument(0) instanceof ImmutableFirstArrayExpr
+      )
+      or
+      exists(Field f |
+        this = f.getAnAccess() and
+        f.isFinal() and
+        f.getInitializer() instanceof ImmutableFirstArrayExpr
+      )
+    )
+  }
+
+  /** Gets the first element of this array. */
+  Expr getFirstElement() {
+    result = this.(Assignment).getRhs().(ImmutableFirstArrayExpr).getFirstElement()
+    or
+    result = this.(LocalVariableDeclExpr).getInit().(ImmutableFirstArrayExpr).getFirstElement()
+    or
+    exists(CommandArgArrayImmutableFirst caa | this = caa.getAUse() |
+      result = caa.getFirstElement()
+    )
+    or
+    result = this.(MethodAccess).getArgument(0).(ImmutableFirstArrayExpr).getFirstElement()
+    or
+    result = this.(FieldAccess).getField().getInitializer()
```

Comment:
Looks like this is missing `.(ImmutableFirstArrayExpr).getFirstElement()`.  Consider perhaps to replace the class `ImmutableFirstArrayExpr` and this member predicate with a single top-level predicate instead.  That way you can avoid essentially repeating the recursion twice.

Reply:
Good point - using a top level predicate would be simpler.